shared-settings-refactor

kghamilton89 opened the pull request:
Shared Settings refactor

rodionos commented:
I opened the PR text in markdown preview mode to view the final, formatted, product and I opened the 5695 rules to see if there are misses. I noticed that some rules are only partially implemented in the proposed version of the md file. I closed the PR without annotating which specific changes to make to bring the document in line with 5695 rules because this work is better be done by the PR author.

kghamilton89 commented:
Each section has been painstakingly reviewed. All possible and default value enumerated. Other 5695 rules followed.

rodionos commented:
See my changes at https://github.com/axibase/charts/pull/86/commits/5004cf6cf2636af69ce51a059ec06c4cda4e9e60. It's not a ready product yet. I create subtables to group settings by subject area, but remove also descriptions and examples.

Because this is a reference document, the requirements are strict and it's important to keep the definitions consistent and concise.

a) Avoid new terms (i,e, 'designation', 'boundary') and generic concepts (visualization, etc). Use specific terms from comp.sci and user interface design.
b) No need to place all possible values on multiple lines, they can be enumerated by separating them with comma.
c) Remove unnecessary examples.
d) Do not forget to mention all settings, e.g. `axis-title-right` was not in the table.
e) Check which settings do not apply to the vast majority of widgets. I'm include to move display-tables, expand tables and perhaps some other out of shared into specific docs.

kghamilton89 commented:
* General re-reading with the diff, I made a few small changes.
* Refactor Possible Values sections.
* I removed almost every case of examples with more than one example. The ones I left demonstrate vastly difference settings, such as using `enabled = false` versus `enabled = value > top(3)` Boolean expression.
* I removed the following settings as I believe they are only used by the Time Chart:
  * `centralize-columns`
  * `markers`
  * `display-panels`
  * `expand-panels`
  * `ahead-time-span`
  * `centralize-ticks`
  * `rotate-ticks`

rodionos commented:
The previous description for these settings said that display-panels and expand-panels are used by Time and Bar charts. rotate-ticks maybe supported by other widgets (box, bar?), pls double check.

kghamilton89 commented:
`display-panels` and `expand-panels` are not supported by Box Chart or Bar Chart, but `rotate-ticks` is. Should we include `rotate-ticks` here even though it is only supported by three widgets?

kghamilton89 commented:
Box and Bar have no vertical ticks (because there is no time axis) so `centralize-ticks` is only supported by Time Chart. My logic was that including both of these settings in Shared may be confusing.

rodionos commented:
Lets move rotate-ticks to each widget doc separately.

kghamilton89 commented:
I think linking to each of the documentation associated with the sub-sections like Rate, Interpolation is a good idea. I did the same thing locally but did so on the header, I'll revert those changes.

rodionos commented:
Take a look a revisions I'm making. Except for formatting changes, the docs are still not precise enough. I'll annotate the remaining changes using code review.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,218 @@
+[`type`](#type) | `type = chart`| Widget visualization type: [`chart`](../time-chart/README.md), [`gauge`](../gauge-chart/README.md), [`bar`](../bar-chart/README.md), [`histogram`](../histogram/README.md), [`box`](../box-chart/README.md), [`calendar`](../calendar-chart/README.md), [`treemap`](../treemap/README.md), [`pie`](../pie-chart/README.md), [`console`](../alert-console/README.md), [`proper ...
+[`tooltip`](#tooltip) | `tooltip = Daily Averages` | Widget description displayed on title mouseover. | [↗](https://apps.axibase.com/chartlab/61c9771e)
+[`header-style`](#header-style)| `header-style = color: red`| CSS style to change the header color or to hide it.| [↗](https://apps.axibase.com/chartlab/b4c45b71/2/)
+[`colors`](#colors) | `colors = green, #ccc` | Comma separated list of colors applied to series shapes: lines, rectangles, or circles, depending on the widget type.<br>Possible values: [color names](https://en.wikipedia.org/wiki/Web_colors) or hex codes.<br>Default value: `steelblue, orange, green, purple`.| [↗](https://apps.axibase.com/chartlab/55dde7a7)
```
We have more than 4 default colors. What are they?

kghamilton89 replied:
Default colors enumerated to the last.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,218 @@
+
+Name | Example | Description | &nbsp;
+:--|:--|:--|:--
+[`timespan`](#timespan) | `timespan = 6 hour` | Date interval specified as the number of [time units](https://axibase.com/docs/atsd/api/data/series/time-unit.html).<br>Format: `count time_unit`.<br>Default value: `1 hour`.| [↗](https://apps.axibase.com/chartlab/f08405c6)
```
Other settings that must be in this section: endtime, start-time, batch-size, batch-interval, other `batch*` settings. Also, timezone.

rodionos replied:
Also missing: multiple-series, update-interval

rodionos replied:
Also missing: url, url-parameters

kghamilton89 replied:
Added:
* `batch-size`
* `batch-update`
* `timezone`
* `multiple-series`
* `update-interval`
* `url`
* `url-parameters`
* `context-path`

kghamilton89 replied:
* `method-path`

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,218 @@
+[`series-limit`](#series-limit) | `series-limit = 10` | Maximum number of series retrieved from the database.<br>Default value: `1000`| [↗](https://apps.axibase.com/chartlab/af34dc29)
+[`limit`](#limit) | `limit = 10` | Maximum number of samples returned for each returned series.<br>Default value: `0` (not limited).|[↗](https://apps.axibase.com/chartlab/13549b9c/2/)<br>[↗](https://apps.axibase.com/chartlab/2604d710)
+[`cache`](#cache) | `cache = true` | Query last values from the cache table for fast response.<br>Default value: `false`| [↗](https://apps.axibase.com/chartlab/032bcfcc)
+[`refresh-interval`](#refresh-interval) | `refresh-interval = 5 minute` | Interval to refresh data specified as the number of [time units](https://axibase.com/docs/atsd/api/data/series/time-unit.html).<br>Format: `count time_unit`.<br>Default value: `5 seconds`. | [↗](https://apps.axibase.com/chartlab/da03b8a5/24/)
```
How is this different from update-interval? `update-interval` is also missing btw.

kghamilton89 replied:
`update-interval` added, and the relationship between the settings described. Chart examples updated as well for further clarity.

kghamilton89 commented:
I've been following along as you make new commits.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,218 @@
+
+Name | Example | Description | &nbsp;
+:--|:--|:--|:--
+[`legend-position`](#legend-position) | `legend-position = left` | Legend location.<br>Possible values: `hidden`, `top`, `right`, `bottom`, `left`<br>Default value: `hidden`.<br>Combine values to define corners.| [↗](https://apps.axibase.com/chartlab/b962dd26)
```
Is it always hidden. What if we request many series using a wildcard?

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,218 @@
+[`axis-title-right`](#axis-title) | `axis-title-right = Free Memory`| Vertical text displayed along the right axis.| [↗](https://apps.axibase.com/chartlab/d8adfe73)
+[`day-format`](#day-format) | `day-format = %y/%m/%d`| [Day format](../../syntax/format-settings.md#day-format) applied to time axis.| [↗](https://apps.axibase.com/chartlab/40dafe6b/2/)
+[`min-range`](#min-range) | `min-range = 0`| Left axis minimum range.<br>If loaded data exceeds `min-range` value, widget is scaled to show true minimum range.| [↗](https://apps.axibase.com/chartlab/d9eeeda5)
+[`max-range`](#max-range) | `max-range = 100`| Left axis maximum range.<br>If loaded data exceeds `max-range` value, widget is scaled to show true maximum range.| [↗](https://apps.axibase.com/chartlab/74052e3e)
```
> widget is scaled to show true maximum range
This is not precise. The terms - 'scaled', 'widget' are too generic too describe the effect of this setting.

kghamilton89 replied:
I rephrased to "chart does (not) display sample minimum/maximum range." I don't know how "chart" could be misunderstood and "display" is the verb we use in nearby `axis-title` and `axis-title-right` settings.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,218 @@
+
+Name | Example | Description | &nbsp;
+:--|:--|:--|:--
+[`rate`](#rate) | `rate = 15 second` | Compute the difference between consecutive samples per unit of time, or [rate period](https://axibase.com/docs/atsd/api/data/series/rate.html#rate-period).<br>Format: `count time_unit` or `auto`.<br>Default value: `auto` (None)| [↗](https://apps.axibase.com/chartlab/d2daeb40)
```
As one useful case, we support `0 time_unit`. Worth mentioning it's effect here and providing a link to chartlab.

kghamilton89 replied:
Mentioned the effect of including `0 time_unit`.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,218 @@
+Name | Example | Description | &nbsp;
+:--|:--|:--|:--
+[`rate`](#rate) | `rate = 15 second` | Compute the difference between consecutive samples per unit of time, or [rate period](https://axibase.com/docs/atsd/api/data/series/rate.html#rate-period).<br>Format: `count time_unit` or `auto`.<br>Default value: `auto` (None)| [↗](https://apps.axibase.com/chartlab/d2daeb40)
+[`rate-counter`](#rate-counter) | `rate-counter = true`| Ignore differences between consecutive samples.<br>Default value: `false`| [↗](https://apps.axibase.com/chartlab/d91e674f)
```
Not clear what it means from this description.

kghamilton89 replied:
expanded to explain that negative time values are converted to `0`

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,218 @@
+Name | Example | Description | &nbsp;
+:--|:--|:--|:--
+[`interpolate-function`](#interpolate-function) | `interpolate-function = linear` | Interpolation function applied to detailed samples.<br>Possible values: `NONE`, `LINEAR`, `PREVIOUS`, `NEXT`, `VALUE(n)`, where `n` is the numerical value to be used to fill missing samples.<br>Default value: `NONE`| [↗](https://apps.axibase.com/chartlab/6221d836)
+[`interpolate-period`](#interpolate-period) | `interpolate-period = 1 minute` | Define the period for interpolated values.<br>Possible values:<br>`count` + [time unit](https://axibase.com/docs/atsd/api/data/series/time-unit.html) | [↗](https://apps.axibase.com/chartlab/6221d836)
```
Refactor this and similar interval settings below to the common description format I specified earlier in the document.

kghamilton89 replied:
refactored this and other interval settings

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,218 @@
+[`offset-right`](#offset) | `offset-right = 50` | Offset from the right page border, in pixels.<br>Default value: `0`. | [↗](https://apps.axibase.com/chartlab/7c30a34b)
+[`offset-top`](#offset) | `offset-top = 50` | Offset from the top page border, in pixels.<br>Default value: `0`. | [↗](https://apps.axibase.com/chartlab/fcbb484b)
+[`offset-bottom`](#offset) | `offset-bottom = 50` | Offset from the bottom page border, in pixels.<br>Default value: `0`. | [↗](https://apps.axibase.com/chartlab/1baa3e3a)
+[`scale`](#scale) | `scale = 0.8` | Chart scale.<br>Value must exceed `0.0`.<br>Default value: `1.0`.| [↗](https://apps.axibase.com/chartlab/1679114f)
```
Missing settings - description, author, title

kghamilton89 replied:
Sasha says `author` and `description` are not settings.

kghamilton89 replied:
Added `title`.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,218 @@
-Confidence Intervals | `value = (1 - lower_confidence(90, 'free') / forecast('total')) * 100`<br>`value = (1 - upper_confidence(90, 'free') / forecast('total')) * 100`| Retrieve upper and lower limits of the confidence interval for the underlying series.<br>Arguments: Level-specified integer from `0` to `100`, and `alias`.| [![](./images/button.png)](https://apps.axibase.com/chartlab/da03b8a5/14/ ...
+Name | Example | Description | &nbsp;
+:--|:--|:--|:--
+[`replace-value`](#replace-value) | `replace-value = value*2`| Modify or filter series values.<br>Supported fields: `value`.| [↗](https://apps.axibase.com/chartlab/1e4dccf0/2/)
```
Specify which fields are accessible in the replace-value expression. It's already documented somewhere.

kghamilton89 replied:
Added list of accessible fields to both.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,218 @@
+:--|:--|:--|:--
+[`replace-value`](#replace-value) | `replace-value = value*2`| Modify or filter series values.<br>Supported fields: `value`.| [↗](https://apps.axibase.com/chartlab/1e4dccf0/2/)
+[`alias`](#alias) | `alias = s1` | Unique series name to pass data to other series. | [↗](https://apps.axibase.com/chartlab/e0623f81)
+[`value`](#value) | `value = max('s1')`| Define series value.<br>Retrieve the value of the underlying series identified by alias.| [↗](https://apps.axibase.com/chartlab/6428aa9c)<br>[↗](https://apps.axibase.com/chartlab/a39c16f4)
```
Specify which fields are accessible in the `value` expression. It's already documented somewhere.

kghamilton89 replied:
Added which fields are accessible for both.

kghamilton89 commented:
What do you think about `sort` setting? At the moment we're placing it in each widget where it is applicable.

rodionos commented:
Yes, it's better to keep `sort` in specific widgets where it is supported and has an effect.

rodionos commented:
This PR is not ready yet. See earlier comments which I submitted as a review.

kghamilton89 commented:
I'm working on those changes right now.

kghamilton89 commented:
@VeselovAlex please take a look at this document and let me know if you have any suggestions for improvements. We should be close to done. All settings included, etc.

kghamilton89 commented:
For `value` and `replace-value` settings, accessible fields are identical. Should we create a separate document called `/accessible-fields.md` for example, to which we can link in each **Description**, as "See [Accessible Fields]() for more information."?
This document can contain an example for each type of accessible field, or at least some of the less obvious ones.
For example, I don't know someone how who is unfamiliar with ChartLab is going to be able to make much use of `threshold_count` or `threshold_duration` without some instruction, as it's not very obvious what that means.
I envision the document as a short **Overview**, followed by a table similar to the one shown below:

Name | Example | Description | &nbsp;
:--|:--|:--|:--
`max` | `value = max('raw')` | Max value for the specified series. | [↗](https://apps.axibase.com/chartlab/2cd7b215)

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,230 @@
+Name | Example | Description | &nbsp;
+:--|:--|:--|:--
+<a name="alias"></a>[`alias`](#alias) | `alias = s1` | Unique series name to pass data to other series. | [↗](https://apps.axibase.com/chartlab/e0623f81)
+<a name="replace-value"></a>[`replace-value`](#replace-value) | `replace-value = value*2`| Modify or filter series values.<br>Supported fields: `metric`, `entity`, `tags`, `value`, `previous`, `movavg`, `detail`, `forecast`, `forecast_deviation`, `lower_confidence`, `upper_confidence`, `percentile`, `max`, `min`, `avg`, `sum`, `delta`, `counter`, `last`, `first`, `min_value_time`, `max_value_time ...
```
I think only 4 fields are accessible in the `replace-value` expression: value, previous value, current sample time, and previous sample time. I'm not sure of the names. Is it true that stat functions are accessible here, for example, avg? Could you check and provide examples?

kghamilton89 replied:
Yes, I was surprised as well: 
* https://apps.axibase.com/chartlab/6bdb20d4
* https://apps.axibase.com/chartlab/78cb418c
* https://apps.axibase.com/chartlab/135a35e4
* https://apps.axibase.com/chartlab/df8ad3e7
* https://apps.axibase.com/chartlab/3e611186

Looks like other fields (such as `metric` and `entity`) only apply when `data-type = forecast` or some similar setting, used to define the series, so I'll remove them.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,230 @@
+:--|:--|:--|:--
+<a name="alias"></a>[`alias`](#alias) | `alias = s1` | Unique series name to pass data to other series. | [↗](https://apps.axibase.com/chartlab/e0623f81)
+<a name="replace-value"></a>[`replace-value`](#replace-value) | `replace-value = value*2`| Modify or filter series values.<br>Supported fields: `metric`, `entity`, `tags`, `value`, `previous`, `movavg`, `detail`, `forecast`, `forecast_deviation`, `lower_confidence`, `upper_confidence`, `percentile`, `max`, `min`, `avg`, `sum`, `delta`, `counter`, `last`, `first`, `min_value_time`, `max_value_time ...
+<a name="value"></a>[`value`](#value) | `value = max('s1')`| Define series value.<br>Retrieve the value of the underlying series identified by alias.<br>Supported fields: `metric`, `entity`, `tags`, `value`, `previous`, `movavg`, `detail`, `forecast`, `forecast_deviation`, `lower_confidence`, `upper_confidence`, `percentile`, `max`, `min`, `avg`, `sum`, `delta`, `counter`, `last`, `first`, `min_v ...
```
Similar to `replace-value`, I think only a) `value('{alias}')` b) statistical functions (no need to list them all, a link will suffice), and c) some built-in functions are accessible. Please check this for correctness.

kghamilton89 replied:
You're correct about replace value, I'll remove other settings.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,230 @@
 ### Derived Value Settings
 
-Specify the `value` setting to create calculated series derived from raw series using arithmetic expressions in JavaScript syntax. The expression returns a number or `null`.
+Specify `value` setting to create calculated series derived from raw series using arithmetic expressions in JavaScript syntax. The expression returns a number or `null`.
```
The expression **must return** a number o `null`. Samples with null value are hidden.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,230 @@
+<a name="min-range"></a>[`min-range`](#min-range) | `min-range = 0`| Left axis minimum range.<br>If loaded data exceeds `min-range` value, chart displays sample minimum range.| [↗](https://apps.axibase.com/chartlab/d9eeeda5)
+<a name="max-range"></a>[`max-range`](#max-range) | `max-range = 100`| Left axis maximum range.<br>If loaded data exceeds `max-range` value, chart displays sample maximum range.| [↗](https://apps.axibase.com/chartlab/74052e3e)
+<a name="min-range-right"></a>[`min-range-right`](#min-range-right) | `min-range-right = 0`| Right axis minimum range.<br>If loaded data exceeds `min-range-right` value, chart displays sample minimum range.| [↗](https://apps.axibase.com/chartlab/a5d7f10b)
+<a name="max-range-right"></a>[`max-range-right`](#max-range-right) | `max-range-right = 100`| Right axis maximum range.<br>If loaded data exceeds `max-range-right` value, chart displays sample maximum range.| [↗](https://apps.axibase.com/chartlab/5f1445c6)
```
> If loaded data exceeds `max-range-right` value, chart displays sample maximum range...

If loaded data exceeds `max-range-right` value, the maximum value displayed on the right axis is adjusted to show the loaded data ... something like that is clear I think, but is somewhat verbose.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,230 @@
+<a name="max-range-force"></a>[`max-range-force`](#max-range-force) | `max-range-force = 100` | Left axis forced minimum and maximum range.<br>If loaded data exceeds `max-range-force`, chart does not display sample maximum range. | [↗](https://apps.axibase.com/chartlab/fa0b58a7)
+<a name="min-range-force"></a>[`min-range-force`](#min-range-force) | `min-range-force = 0` | Left axis forced minimum and maximum range.<br>If loaded data exceeds `min-range-force`, chart does not display sample minimum range. | [↗](https://apps.axibase.com/chartlab/fa0b58a7)
+<a name="min-range-right-force"></a>[`min-range-right-force`](#min-range-right-force) | `min-range-right-force = 0` | Right axis forced minimum range.<br>If loaded data exceeds `min-range-right-force`, chart does not display sample maximum range. | [↗](https://apps.axibase.com/chartlab/ac40afc3)
+<a name="max-range-right-force"></a>[`max-range-right-force`](#max-range-right-force) | `max-range-right-force = 100` | Right axis forced maximum range.<br>If loaded data exceeds `max-range-right-force`, chart does not display sample maximum range. | [↗](https://apps.axibase.com/chartlab/ac40afc3)
```
> If loaded data exceeds `max-range-right-force`, chart does not display sample maximum range.

This is hard to read. What happens is that the maximum value displayed on the right axis remains set even if the loaded data contains samples that exceed `max-range-right-force`. Such samples are hidden. Rephrase for  clarity.

kghamilton89 replied:
Rephrased with the negation of above sentence: "axis range is not adjusted to show loaded data"

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,230 @@
+
+Name | Example | Description | &nbsp;
+:--|:--|:--|:--
+<a name="timespan"></a>[`timespan`](#timespan) | `timespan = 6 hour` | Date interval specified as the number of [time units](https://axibase.com/docs/atsd/api/data/series/time-unit.html).<br>When specified without `start-time` or `endtime` setting, `endtime` is `now`.<br>Format: `count time_unit`.<br>Default value: `1 hour`.| [↗](https://apps.axibase.com/chartlab/f08405c6)
```
I'd say it's 'Data selection interval', not Date... If end-time and start time are not set, the data is loaded for an interval that starts with current time minus timespan, and ends with current time.

Also, when referring to other settings, keep names consistent, and use hyphens, e.g. `end-time`, `start-time`, but not `endtime`.

kghamilton89 replied:
Changed

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,230 @@
+<a name="title"></a>[`title`](#title) | `title = CPU Usage Statistics` | Widget title, displayed above chart.<br>HTML Markup is supported. | [↗](https://apps.axibase.com/chartlab/6f3fc5f0)
+<a name="tooltip"></a>[`tooltip`](#tooltip) | `tooltip = Daily Averages` | Widget description displayed on title mouseover. | [↗](https://apps.axibase.com/chartlab/61c9771e)
+<a name="header-style"></a>[`header-style`](#header-style)| `header-style = color: red`| CSS style to change the header color or to hide it.| [↗](https://apps.axibase.com/chartlab/b4c45b71/2/)
+<a name="colors"></a>[`colors`](#colors) | `colors = green, #ccc` | Comma separated list of colors applied to series shapes: lines, rectangles, or circles, depending on the widget type.<br>Possible values: [color names](https://en.wikipedia.org/wiki/Web_colors) or hex codes.<br>Default values: `steelblue, orange, green, purple`, `maroon`, `yellowgreen`, `hotpink`, `chocolate`, `deepskyblue`.<br>A ...
```
Review this and other examples. Remove any setting that doesn't relate directly to the setting being demoed or doesn't help to show it. For example, `colors` should not be really just one color. Otherwise, one can assign color to series using `color` setting for each series separately.

Here's a better example:

* All non-relevant settings removed. Inheritance for metric applied to avoid duplication
* Colors is a list. colors are different enough to be of contrast (not shades of the same color).

```
[configuration]
  height-units = 2
  width-units = 2
  

[group]

  [widget]
    type = chart
    timespan = 15 minute
    colors = green, orange
    
    metric = cpu_busy

    [series]
      entity = nurswgvml007

    [series]
      entity = nurswgvml006
```

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,230 @@
+:--|:--|:--|:--
+<a name="timespan"></a>[`timespan`](#timespan) | `timespan = 6 hour` | Date interval specified as the number of [time units](https://axibase.com/docs/atsd/api/data/series/time-unit.html).<br>When specified without `start-time` or `endtime` setting, `endtime` is `now`.<br>Format: `count time_unit`.<br>Default value: `1 hour`.| [↗](https://apps.axibase.com/chartlab/f08405c6)
+<a name="start-time"></a>[`start-time`](#start-time)| `start-time = 2018-07-05 12:00:00` | Start time of interval specified in [ISO format](https://axibase.com/docs/atsd/shared/date-format.html).<br>When `start-time` is specified, either `endtime` **or** `timespan` is also required. | [↗](https://apps.axibase.com/chartlab/451287d3)
+<a name="end-time"></a>[`end-time`](#end-time) | `endtime = 2018-07-05 13:00:00` | End time of interval specified in [ISO format](https://axibase.com/docs/atsd/shared/date-format.html).<br>When `endtime` is specified, either `start-time` **or** `timespan` is also required.| [↗](https://apps.axibase.com/chartlab/1565799f)
```
Definitions for endtime and starttime are not correct. They support a) calendar keywords and expressions b) ISO dates c) local dates. Date/time pattern is `specify here`.

kghamilton89 replied:
Added calendar syntax and local dates, as well as link to calendar syntax documentation.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,230 @@
+<a name="start-time"></a>[`start-time`](#start-time)| `start-time = 2018-07-05 12:00:00` | Start time of interval specified in [ISO format](https://axibase.com/docs/atsd/shared/date-format.html).<br>When `start-time` is specified, either `endtime` **or** `timespan` is also required. | [↗](https://apps.axibase.com/chartlab/451287d3)
+<a name="end-time"></a>[`end-time`](#end-time) | `endtime = 2018-07-05 13:00:00` | End time of interval specified in [ISO format](https://axibase.com/docs/atsd/shared/date-format.html).<br>When `endtime` is specified, either `start-time` **or** `timespan` is also required.| [↗](https://apps.axibase.com/chartlab/1565799f)
+<a name="url"></a>[`url`](#url) | `url = https://atsd.example.org:8443` | URL of the ATSD server.<br>The setting is necessary if the data is loaded from an ATSD server running on a different host. | [↗](https://apps.axibase.com/chartlab/88885921)
+<a name="context-path"></a>[`context-path`](#context-path) | `context-path = /api/v2/`| Context path is the prefix of the URL path that is used to select the context to which an incoming request is passed.<br>Default value: `/api/v1/`.|[↗](https://apps.axibase.com/chartlab/ed175a7b)
```
When you say incoming request, who is it incoming too? The words assume a place where it's incoming, wherE?

kghamilton89 replied:
Specified as, "to which an incoming request to ATSD server is directed." Is this correct?

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,230 @@
+<a name="series-limit"></a>[`series-limit`](#series-limit) | `series-limit = 10` | Maximum number of series retrieved from the database.<br>Default value: `1000`| [↗](https://apps.axibase.com/chartlab/af34dc29)
+<a name="limit"></a>[`limit`](#limit) | `limit = 10` | Maximum number of samples returned for each returned series.<br>Default value: `0` (not limited).|[↗](https://apps.axibase.com/chartlab/13549b9c/2/)<br>[↗](https://apps.axibase.com/chartlab/2604d710)
+<a name="cache"></a>[`cache`](#cache) | `cache = true` | Query last values from the cache table for fast response.<br>Default value: `false`.| [↗](https://apps.axibase.com/chartlab/032bcfcc)
+<a name="update-interval"></a>[`update-interval`](#update-interval) | `update-interval = 5 min` | Interval to update widget specified as the number of [time units](https://axibase.com/docs/atsd/api/data/series/time-unit.html).<br>`update-interval` can contain several `refresh-interval` periods, whereby data is updated in memory several times but only added to the widget each specified interval.<b ...
```
Is `min` a supported time unit? It should not be. Also, I think unit is singular, and our validator in VScode will even throw an error. Please re-check ALL settings in this file with our VScode validator.

kghamilton89 replied:
This is a typo.

rodionos commented:
I just realized that we cannot have `offset-*` settings in this document, because they apply to the entire portal, and hence to the `[configuration]` section. Move these settings there, and I would group them into a `Layout` table in the target document, along with `width-units`, `height-units`, and `widgets-per-row` settings. 

The layout settings that apply to widgets are: `top-units`, `left-units`, `width-units`, `height-units`.

kghamilton89 commented:
It's not clear whether you want these settings to be in a new document altogether or just moved into a separate table in this document.

rodionos commented:
New file unless we already have a file that covers layout. This file deals with shared WIDGET settings so we cannot keep these offset settings in this file.

kghamilton89 commented:
Understood. We have a `/configuration` directory. I'll move these and other `[configuration]` settings there. `/configuration/settings.md`.

rodionos commented:
Lets copy what's relevant from https://github.com/axibase/atsd/blob/master/portals/portal-settings.md which covers layout. For some reason, we don't have these settings in charts.

kghamilton89 commented:
Oh that's convenient. Alright. I'll make that document.

kghamilton89 commented:
* All settings checked against source documentation.
* All charts verified by Charts plugin.
* Extra settings removed where possible, target settings isolated.
* "Boolean Expression" terminology removed where appropriate, replaced with "Possible values: `false`, `true`.

kghamilton89 commented:
Build is failing because of live links read as dead.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,244 @@
+<a name="update-interval"></a>[`update-interval`](#update-interval) | `update-interval = 5 minute` | Data update interval specified as the number of [time units](https://axibase.com/docs/atsd/api/data/series/time-unit.html).<br>Format: `count time_unit`.<br>Default value: `1 minute`. | [↗](https://apps.axibase.com/chartlab/724a08aa)
+<a name="batch-update"></a>[`batch-update`](#batch-update) | `batch-update = true`| Send data queries to the server in batches with size specified in `batch-size` setting.<br>Possible values: `false`, `true`.<br>If `true`, series for which the request has failed are requested separately from successfully updated series.<br>Default value: `false`.| [↗](https://apps.axibase.com/chartlab/669ac522)
+<a name="batch-size"></a>[`batch-size`](#batch-size) | `batch-size = 1` | Maximum number of series per server batch request.<br>If `0` is specified, the limit is not set.<br>Valid when `batch-update = true`.<br>Default value: `8`.| [↗](https://apps.axibase.com/chartlab/4dda9f75)
+<a name="refresh-interval"></a>[`refresh-interval`](#refresh-interval) | `refresh-interval = 5 minute` | Data refresh interval specified as the number of [time units](https://axibase.com/docs/atsd/api/data/series/time-unit.html).<br>If `update-interval` is not specified, `update-interval = refresh-interval`.<br>Format: `count time_unit`.<br>Default value: `5 seconds`. | [↗](https://apps.axibase.c ...
```
Refresh settings are incorrectly described. See https://axibase.com/products/axibase-time-series-database/visualization/widgets/configuring-the-widgets/. Revert to include correct descriptions. Other data update settings except update-interval are probably also affected.

kghamilton89 replied:
I believe there are errors in the original documentation. The source document says for example, "The time interval, during which there is no update (0 by default) from the time of receiving the last values of the series. Refresh-interval is set in seconds." 

But actually, this setting supports `count time_units` input: https://apps.axibase.com/chartlab/bbf8dec0.

Likewise, when no `refresh-interval` is set, `update-interval = refresh-interval`: https://apps.axibase.com/chartlab/2657e1b9

rodionos replied:
![image](https://user-images.githubusercontent.com/2098022/46594199-18a5cb00-cada-11e8-8d1e-caf214c2c122.png)

https://apps.axibase.com/chartlab/2657e1b9/2/

As you can see, no update requests are made every 5 seconds.

kghamilton89 replied:
The same is true of `retry-refresh-interval`, `count time_units` is accepted: https://apps.axibase.com/chartlab/d7e21c29

The same is true of `error-refresh-interval`: https://apps.axibase.com/chartlab/48e8b5cc

kghamilton89 replied:
So the idea is that update-interval actively queries the server, whereas all `refresh-*` settings check to see if the server has sent new information?

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,244 @@
+<a name="retry-refresh-interval"></a>[`retry-refresh-interval`](#retry-refresh-interval) | `retry-refresh-interval = 5 minute`| Interval to skip data requests for the given series if the previous request received empty response, specified as the number of [time units](https://axibase.com/docs/atsd/api/data/series/time-unit.html).<br>Format: `count time_unit`. | [↗](https://apps.axibase.com/chartl ...
+<a name="error-refresh-interval"></a>[`error-refresh-interval`](#error-refresh-interval) | `error-refresh-interval = 30 minute`| Interval to skip data requests for the given series if the previous request failed, specified as the number of [time units](https://axibase.com/docs/atsd/api/data/series/time-unit.html).<br>Format: `count time_unit`. | [↗](https://apps.axibase.com/chartlab/48e8b5cc)
+
+### Data Source
```
Add `see also` linking to https://axibase.com/docs/atsd/api/data/#data-api-endpoints

kghamilton89 replied:
Link added.

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,244 @@
+<a name="table"></a>[`table`](#table) | `table = KLZ_CPU`| Table in the relational database from which to retrieve numeric values.<br>Alternative to `metric` setting.<br>Both `table` and `attribute` must be defined.| [↗](https://apps.axibase.com/chartlab/35fde2bf)
+<a name="attribute"></a>[`attribute`](#attribute) | `attribute = Current_Average` | Column name in a relational database table. The column must be of numeric data type.  | [↗](https://apps.axibase.com/chartlab/35fde2bf)
+<a name="data-type"></a>[`data-type`](#data-type) | `data-type = forecast` | Series data type.<br>Possible values: `history`, `forecast`, `forecast_deviation`, `lower_confidence`, `upper_confidence`.| [↗](https://apps.axibase.com/chartlab/2942e708)
+<a name="forecast-name"></a>[`forecast-name`](#forecast-name) | `forecast-name = hw5` | Forecast name.<br>If no forecast name is defined, [default series forecast](https://axibase.com/docs/atsd/forecasting/#persistence-settings) is loaded.| [↗](https://apps.axibase.com/chartlab/533918cc)
```
https://axibase.com/docs/atsd/api/data/series/query.html#forecast-filter is a useful link here I think

rodionos commented on widgets/shared/README.md:
```diff
@@ -1,171 +1,244 @@
+<a name="data-type"></a>[`data-type`](#data-type) | `data-type = forecast` | Series data type.<br>Possible values: `history`, `forecast`, `forecast_deviation`, `lower_confidence`, `upper_confidence`.| [↗](https://apps.axibase.com/chartlab/2942e708)
+<a name="forecast-name"></a>[`forecast-name`](#forecast-name) | `forecast-name = hw5` | Forecast name.<br>If no forecast name is defined, [default series forecast](https://axibase.com/docs/atsd/forecasting/#persistence-settings) is loaded.| [↗](https://apps.axibase.com/chartlab/533918cc)
+
+### Entity Filter
```
The links in this section are strange. As if we don't have any REST API docs covering entity filtering. But the fact is - we do have docs: https://axibase.com/docs/atsd/api/data/filter-entity.html#entity-filter-fields and https://axibase.com/docs/atsd/api/data/series/query.html#entity-filter
